models/track_and_reid_model.py

```diff
@@ -0,0 +1,168 @@
+    parser.add_argument('--output', help='output video file (mp4 format) or folder')
+    parser.add_argument('--track_checkpoint', help='checkpoint file for the track model')
+    parser.add_argument('--reid_checkpoint', help='checkpoint file for the reID model')
+    parser.add_argument('--device', default='cuda:0', help='device used for inference')
+    parser.add_argument('--show', action='store_true', help='whether show the results on the fly')
+    parser.add_argument('--backend', choices=['cv2', 'plt'], default='cv2', help='the backend to visualize the results')
+    parser.add_argument('--fps', help='FPS of the output video')
+    parser.add_argument('--crops_folder', help='Path to the folder in which the generated crops should be saved')
+    parser.add_argument("--reid_opts", help="Modify reid-config options using the command-line 'KEY VALUE' pairs", default=[], nargs=REMAINDER,)
+    args = parser.parse_args()
+    return args
+
+
+def set_reid_cfgs(args):
+    cfg = get_cfg()
+    cfg.merge_from_file(args.reid_config)
+    cfg.merge_from_list(args.reid_opts)
+    cfg.freeze()
+    return cfg
+
+
+def apply_reid_model(reid_model, data):
+    feats = []
+    pids = []
+    camids = []
+    print('Converting test data to feature vectors:')
+    # Converts all images in the bounding_box_test and query to feature vectors
+    for (feat, pid, camid) in tqdm.tqdm(reid_model.run_on_loader(data), total=len(data)):
+        feats.append(feat)
+        pids.extend(pid)
+        camids.extend(camid)
+
+    print("The size of test gallery is", len(pids))
+    feats = torch.cat(feats, dim=0)
+    g_feat = feats
+    g_pids = np.asarray(pids)
+    # g_camids = np.asarray(camids)  # todo: remove?
+    return feats, g_feat, g_pids
+
+
+def find_best_match(q_feat, g_feat, g_pids):
+    """
+    Given feature vectors of the query images, return the ids of the images that are most similar in the test gallery
+    """
+    features = F.normalize(q_feat, p=2, dim=1)
+    others = F.normalize(g_feat, p=2, dim=1)
+    distmat = 1 - torch.mm(features, others.t())
+
+    distmat = distmat.numpy()
+    best_match_in_gallery = np.argmin(distmat, axis=1)
+    return g_pids[best_match_in_gallery]
+
+
+def tracking_inference(tracking_model, img, frame_id, acc_threshold=0.8):
+    result = inference_mot(tracking_model, img, frame_id=frame_id)
+    # result['track_bboxes'] = result.get('track_results')
+    acc = result['track_results'][0][:, -1]
+    mask = np.where(acc > acc_threshold)
+    result['track_results'][0] = result['track_results'][0][mask]
+    result['bbox_results'][0] = result['bbox_results'][0][mask]
+    return result
+
+
+def reid_inference(reid_model, img, result, frame_id, crops_folder=None):
+    crops_bboxes = result['track_results'][0][:, 1:-1]
+    crops_imgs = mmcv.image.imcrop(img, crops_bboxes, scale=1.0, pad_fill=None)
+    q_feat = torch.empty((len(crops_imgs), 2048))
```

Comment:
note the 2048 here, this is exactly the size of the feature vector. we should have something a bit more stable if we will like to change the architecture of the fast-re-id. I would put crop_images[0].shape[1] (should give the size of the feature vector)

Reply:
As you said it is the size of the feature vector, `crop_images` holds the different crops we extract from the image and its shape is the shape of every crop and not of the feature vector. There is no easy way to know here what the size of the feature vector is, but I don't think it is that terrible - once we add new features, this will be the only place we need to change the dimensions